### PR TITLE
server: (router) add LRU scheduler

### LINKED ISSUE #21678: Misc. bug: Router - Queue model loading requests instead of unloading models with active queries
### Name and Version

llama-server --version
version: 8736 (c8ac02fa1)
built with GNU 13.3.0 for Linux x86_64

### Operating systems

Linux

### Which llama.cpp modules do you know to be affected?

llama-server

### Command line

```shell
CUDA_VISIBLE_DEVICES=1,0,2 CUDA_SCALE_LAUNCH_QUEUES=4x llama-server   --models-preset /mnt/fast1/gguf-models/models-build.ini   -t 6   --host 0.0.0.0   --port 8080   --models-max 1   -fa auto   --log-colors on   -cram 32768   --slot-save-path /mnt/fast1/kv-cache-storage   --cache-prompt   --jinja
```

### Problem description & steps to reproduce

Problem description
When models_max is reached in the router server, the LRU unloading logic (unload_lru) removes the least recently used model to make room for a new one without checking if that model is currently processing active requests. This causes ongoing queries to be interrupted when a request for a different model triggers an immediate unload of the active model.
Steps to reproduce
1. Start the router server with models_max=1 and at least two available models (Model A and Model B).
2. Send a long-running query (e.g., a large generation) to Model A.
3. While Model A is still generating, send a query to Model B.
4. Observe that the server immediately unloads Model A to load Model B, interrupting the first query.

### First Bad Commit

Unknown

### Relevant log output

<details>
<summary>Logs</summary>

```console
[56021] main: model loaded
[56021] main: server is listening on http://127.0.0.1:56021
[56021] main: starting the main loop...
[56021] cmd_child_to_router:ready
srv  proxy_reques: proxying request to model Gemma-4-Pro on port 56021
[56021] srv  update_slots: all slots are idle
[56021] srv    operator(): child server monitoring thread started, waiting for EOF on stdin...
[56021] srv  params_from_: Chat format: peg-gemma4
[56021] slot get_availabl: id  1 | task -1 | selected slot by LRU, t_last = -1
[56021] srv  get_availabl: updating prompt cache
[56021] srv          load:  - looking for better prompt, base f_keep = -1.000, sim = 0.000
[56021] srv        update:  - cache state: 0 prompts, 0.000 MiB (limits: 32768.000 MiB, 262144 tokens, 34359738368 est)
[56021] srv  get_availabl: prompt cache update took 0.00 ms
[56021] slot launch_slot_: id  1 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> min-p -> ?xtc -> ?temp-ext -> dist 
[56021] slot launch_slot_: id  1 | task 0 | processing task, is_child = 0
[56021] slot update_slots: id  1 | task 0 | new prompt, n_ctx_slot = 262144, n_keep = 0, task.n_tokens = 82470
[56021] slot update_slots: id  1 | task 0 | n_tokens = 0, memory_seq_rm [0, end)
[56021] slot update_slots: id  1 | task 0 | prompt processing progress, n_tokens = 8192, batch.n_tokens = 8192, progress = 0.099333
srv  log_server_r: done request: GET / 172.23.0.2 200
srv  log_server_r: done request: GET / 192.168.129.119 200
srv  proxy_reques: proxying request to model Gemma-4-Pro on port 56021
[56021] slot update_slots: id  1 | task 0 | n_tokens = 8192, memory_seq_rm [8192, end)
[56021] slot update_slots: id  1 | task 0 | prompt processing progress, n_tokens = 16384, batch.n_tokens = 8192, progress = 0.198666
srv  log_server_r: done request: GET / 172.23.0.2 200
srv  log_server_r: done request: GET / 192.168.129.119 200
[56021] slot update_slots: id  1 | task 0 | n_tokens = 16384, memory_seq_rm [16384, end)
[56021] slot update_slots: id  1 | task 0 | 16384 tokens since last checkpoint at 0, creating new checkpoint during processing at position 24576
[56021] slot update_slots: id  1 | task 0 | prompt processing progress, n_tokens = 24576, batch.n_tokens = 8192, progress = 0.297999
[56021] slot update_slots: id  1 | task 0 | created context checkpoint 1 of 8 (pos_min = 13824, pos_max = 16383, n_tokens = 16384, size = 2000.031 MiB)
srv    unload_lru: models_max limit reached, removing LRU name=Gemma-4-Pro
srv        unload: stopping model instance name=Gemma-4-Pro
srv    operator(): stopping model instance name=Gemma-4-Pro
[56021] srv    operator(): exit command received, exiting...
srv  log_server_r: done request: GET / 172.23.0.2 200
srv  log_server_r: done request: GET / 192.168.129.119 200
[56021] srv    operator(): operator(): cleaning up before exit...
srv  log_server_r: done request: GET / 172.23.0.2 200
srv  log_server_r: done request: GET / 192.168.129.119 200
srv    operator(): force-killing model instance name=Gemma-4-Pro after 10 seconds timeout
srv    operator(): http client error: Failed to read connection
srv  log_server_r: done request: POST /v1/chat/completions 192.168.129.119 500
srv  log_server_r: done request: GET / 172.23.0.2 200
srv    operator(): instance name=Gemma-4-Pro exited with status 1
srv          load: spawning server instance with name=Gemma-4-Flash on port 41753
```
</details>

### PR DESCRIPTION
**Pull request by ngxson**
## Overview

Add LRU scheduler that queue the eviction requests

If server capacity runs out, `ensure_model_ready()` spawn a wait loop until the requested model is loaded. All queued requests waiting for same model will be unblocked at the same time.

If waiting req suddenly disconnects, it will clear itself from LRU queue

Fix https://github.com/ggml-org/llama.cpp/issues/21678

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: initial idea was made by me, AI used for iterating from that

### CODE REVIEW

**Comment by ngxson** [comment]
@ggerganov this will prevent the problem where "loading" model is evicted that you reported the other day, please give it a try if you can, thanks!

File: tools/server/server-models.h [hunk 1]
```diff
@@ -107,9 +107,11 @@ struct server_model_meta {
 
 struct server_models_routes;
 struct server_subproc; // defined in server-models.cpp
+struct lru_sched;      // defined in server-models.cpp
```
Comment (ggerganov) [inline tools/server/server-models.h]: Naming this `server_lru_sched` would be more consistent

File: tools/server/server-models.cpp [hunk 2]
```diff
@@ -241,8 +424,11 @@ server_models::server_models(
         LOG_WRN("using original argv[0] as fallback: %s\n", argv[0]);
     }
     load_models();
+    debug_fake_timing = !common_get_env("LLAMA_DEBUG_FAKE_TIMING").empty();
```
Comment (ggerganov) [inline tools/server/server-models.cpp]: Think we should unify all server-related env vars to start with the `LLAMA_SERVER_` prefix:

```suggestion
    debug_fake_timing = !common_get_env("LLAMA_SERVER_DEBUG_FAKE_TIMING").empty();
```

**Review by ggerganov — APPROVED** [review]
Couple of naming nits. Confirmed working on my setup now.

**Comment by ngxson** [comment]
@ggerganov can you reapprove this?

**Review by ggerganov — APPROVED** [review]

**Comment by ngxson** [comment]
still need 2nd approval 😓 cc @ServeurpersoCom or @allozaur if you can approve it

better not to use the stacked PR at this stage, it's still too experimental

**Review by allozaur — APPROVED** [review]
